### PR TITLE
Revisit deletion policy after release the last snapshot

### PR DESCRIPTION
**Pull request by dnhatn**
A follow-up of #28140 (see https://github.com/elastic/elasticsearch/pull/28140#discussion_r162458207)

We currently revisit the index deletion policy whenever the global checkpoint has advanced enough. We should also revisit the deletion policy after releasing the last snapshot of a snapshotting commit. With this change, the old index commits will be cleaned up as soon as possible.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java [hunk 1]
```diff
@@ -235,7 +239,7 @@ private static int indexOfKeptCommits(List<? extends IndexCommit> commits, long
      */
     boolean hasUnreferencedCommits() throws IOException {
         final IndexCommit lastCommit = this.lastCommit;
-        if (safeCommit != lastCommit) { // Race condition can happen but harmless
+        if (safeCommit != lastCommit && pendingSnapshots == 0) { // Race condition can happen but harmless
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: I don't really follow this logic? safe commit is maintained outside of the real of snapshots? shouldn't these be two different things? also, we now call this method when the translog is synced and thus potentially something changed for the safe commit. That doesn't have much todo with snapshots and the chance of them being released. Maybe we should make release commit return a boolean and use that an indication that there are commits that can be cleaned up?

**Comment by dnhatn** [comment]
@bleskes I've updated the PR according to our discussion. Can please have another look? Thank you!

File: server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java [hunk 2]
```diff
@@ -178,6 +180,8 @@ synchronized void releaseCommit(final IndexCommit snapshotCommit) {
         if (refCount == 0) {
             snapshottedCommits.remove(releasingCommit);
         }
+        // The commit can be clean up only if no pending snapshot and it is not the last commit.
+        return refCount == 0 && releasingCommit.equals(lastCommit) == false;
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: don't we need a similar check for the safe commit?

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 3]
```diff
@@ -4449,6 +4447,40 @@ public void testCleanUpCommitsWhenGlobalCheckpointAdvanced() throws Exception {
...
+            for (int i = 0; i < numSnapshots; i++) {
+                snapshots.add(engine.acquireIndexCommit(true, false)); // taking snapshots from the safe commit.
+            }
+            // Global checkpoint advanced - clean up all commits except the last commit and the safe commit (snapshotted).
+            globalCheckpoint.set(randomLongBetween(engine.getLocalCheckpointTracker().getCheckpoint(), Long.MAX_VALUE));
+            engine.syncTranslog();
+            assertThat(DirectoryReader.listCommits(store.directory()), contains(safeCommit, lastCommit));
+            for (int i = 0; i < numSnapshots - 1; i++) {
+                snapshots.get(i).close();
+                assertThat(DirectoryReader.listCommits(store.directory()), contains(safeCommit, lastCommit));
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I think we need to check nothing is cleaned just yet?

**Review by bleskes — CHANGES_REQUESTED** [review]
Left some  minor questions. 

@s1monw wdyt about this?

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 3 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I've updated the assertion.

File: server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: Yes, it should return false if the releasing snapshot is the safe commit.

**Review by s1monw — APPROVED** [review]
I am good with this.

**Comment by dnhatn** [comment]
Thanks @bleskes and @s1monw for reviewing!

**Pull request merged**